README.md

```diff
@@ -0,0 +1,226 @@
+
+## xtm_queue_pop_ptrs
+
+Function gets up to count elements from the queue and saves them in pointer array.
+If producer thread pushes ptrs with `XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS`
+flag, user should retrieve and reset "producer failed to put an item
+in the queue and expects notification" flag using `xtm_queue_get_reset_was_full`.
+If this flag was true, user should notify producer thread (see simple example further).
+Return count of extracted messages.
+
+## xtm_queue_consume
+
+Function reads from internal queue pipe, according to file descriptor passed
+to this function.
+Return 0 on success. Otherwise -1 with errno set appropriately (as in `read(2)`,
+since it implies read from internal fd).
+
+## xtm_queue_get_reset_was_full
+
+Function retrieves and resets "producer failed to put an item in the queue and
+expects notification" flag.
+
+Examples
+--------
+
+Full example you can see in test or perf folders, here is only simple
+example of using library.
+
+** Allocate xtm queue **
+
+```c
+xtm_queue = xtm_queue_new(XTM_QUEUE_SIZE);
+```
+
+** Get internal fd on consumer side and pool it in the loop **
+
+```c
+int fd = xtm_queue_consumer_fd(xtm_queue);
+struct pollfd pfds[1];
+pfds[0].fd = fd;
+pfds[0].events = POLLIN;
+int rc;
+while ((rc = poll(pfds, 1, -1)) < 0 && errno == EINTR)
+	;
+if (rc <= 0 || (pfds[0].revents & POLLIN) == 0)
+	//error handling
+```
+
+** Invoke functions in case when producer thread push functions **
+
+```c
+xtm_queue_invoke_funs_all(xtm_queue);
+```
+
+** Pop pointers to data in case when producer thread push pointers **
+
+```c
+unsigned count = xtm_queue_count(xtm_queue);
+unsigned cnt = 0;
+while (cnt < count) {
+	void *ptr_array[BATCH_COUNT_MAX];
+	unsigned rc = xtm_queue_pop_ptrs(xtm_queue, ptr_array,
+					 BATCH_COUNT_MAX);
+	for (unsigned i = 0; i < rc; i++)
+		// do something with ptr_array[i]
+	cnt += rc;
+}
+/* Try to notify producer again if queue was full, if it's fails - panic */
+if (xtm_queue_get_reset_was_full(xtm_queue) &&
+    xtm_queue_notify_producer(xtm_queue) != 0)
+	panic();
+```
+
+** Delete xtm queue, when it is no longer needed, close all internal fds **
+
+```c
+unsigned flags = 0;
+flags |= XTM_QUEUE_MUST_CLOSE_PRODUCER_READFD |
+         XTM_QUEUE_MUST_CLOSE_CONSUMER_READFD;
+xtm_queue_delete(xtm_queue, flags);
+```
+
+** Get internal fd on producer side, push fun to consumer thread **
+
+```c
+unsigned flags = XTM_QUEUE_PRODUCER_NEEDS_NOTIFICATIONS;
+int fd = xtm_queue_producer_fd(xtm_queue);
+if (xtm_queue_push_fun(xtm_queue, function, pointer, flags) < 0) {
+try_next:
+	struct pollfd pfds[1];
+	pfds[0].fd = fd;
+	pfds[0].events = POLLIN;
+	int rc;
+	// Wait for consumer notification, that queue is not full
+	while ((rc = poll(pfds, 1, -1)) < 0 && errno == EINTR)
+		;
+	if (rc <= 0 || (pfds[0].revents & POLLIN) == 0)
+		//error handling
+	xtm_queue_consume(fd);
+	if (xtm_queue_push_fun(xtm_queue, function, pointer, flags) < 0)
```

Comment:
You should just `goto` from here to the earlier call of `xtm_queue_push_fun()` instead of copy-paste.
There are below.